google/cloud/internal/async_streaming_write_rpc_impl.h

```diff
@@ -0,0 +1,208 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_ASYNC_STREAMING_WRITE_RPC_IMPL_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_ASYNC_STREAMING_WRITE_RPC_IMPL_H
+
+#include "google/cloud/completion_queue.h"
+#include "google/cloud/grpc_error_delegate.h"
+#include "google/cloud/internal/async_streaming_write_rpc.h"
+#include "google/cloud/internal/completion_queue_impl.h"
+#include "google/cloud/options.h"
+#include "google/cloud/version.h"
+#include "absl/functional/function_ref.h"
+#include "absl/types/optional.h"
+#include <grpcpp/support/async_stream.h>
+#include <memory>
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace internal {
+
+/**
+ * Wrapper for Asynchronous Streaming Write RPCs.
+ *
+ * A wrapper for gRPC's asynchronous streaming write APIs, which can be
+ * combined with `google::cloud::CompletionQueue` and `google::cloud::future<>`
+ * to provide easier-to-use abstractions than the objects returned by gRPC.
+ */
+template <typename Request, typename Response>
+class AsyncStreamingWriteRpcImpl
+    : public AsyncStreamingWriteRpc<Request, Response> {
+ public:
+  AsyncStreamingWriteRpcImpl(
+      std::shared_ptr<CompletionQueueImpl> cq,
+      std::unique_ptr<grpc::ClientContext> context,
+      std::unique_ptr<Response> response,
+      std::unique_ptr<grpc::ClientAsyncWriterInterface<Request>> stream)
+      : cq_(std::move(cq)),
+        context_(std::move(context)),
+        response_(std::move(response)),
+        stream_(std::move(stream)) {}
+
+  void Cancel() override { context_->TryCancel(); }
+
+  future<bool> Start() override {
+    struct OnStart : public AsyncGrpcOperation {
+      promise<bool> p;
+      Options options = CurrentOptions();
+      bool Notify(bool ok) override {
+        OptionsSpan span(options);
+        p.set_value(ok);
+        return true;
+      }
+      void Cancel() override {}
+    };
+    auto op = std::make_shared<OnStart>();
+    cq_->StartOperation(op, [&](void* tag) { stream_->StartCall(tag); });
+    return op->p.get_future();
+  }
+
+  future<bool> Write(Request const& request,
+                     grpc::WriteOptions options) override {
```

Comment:
optional: maybe change the parameter to `write_options` so we can avoid the need for `options_` in the implementation? or s/options_/opts/